app/src/main/java/de/westnordost/streetcomplete/data/osm/upload/SplitSingleWayUpload.kt

```diff
@@ -0,0 +1,301 @@
+package de.westnordost.streetcomplete.data.osm.upload
+
+import de.westnordost.osmapi.common.errors.OsmConflictException
+import de.westnordost.osmapi.map.MapDataDao
+import de.westnordost.osmapi.map.data.*
+import javax.inject.Inject
+import de.westnordost.osmapi.map.data.Element.Type.*
+import de.westnordost.streetcomplete.data.osm.changes.SplitAtLinePosition
+import de.westnordost.streetcomplete.data.osm.changes.SplitAtPoint
+import de.westnordost.streetcomplete.data.osm.changes.SplitPolylineAtPosition
+import de.westnordost.streetcomplete.ktx.*
+import de.westnordost.streetcomplete.util.SphericalEarthMath
+import kotlin.math.sign
+
+/** Uploads one split way
+ *  Returns only the ways that have been updated or throws a ConflictException */
+class SplitSingleWayUpload @Inject constructor(private val osmDao: MapDataDao)  {
+
+    fun upload(changesetId: Long, way: Way, splits: List<SplitPolylineAtPosition>): List<Way> {
+        val updatedWay = way.fetchUpdated()
+            ?: throw ElementDeletedException("Way #${way.id} has been deleted")
+        if(updatedWay.isClosed() && splits.size < 2)
+            throw ElementConflictException("Must specify at least two split positions for a closed way")
+        checkForConflicts(way, updatedWay)
+
+        val nodes = updatedWay.fetchNodes()
+        val positions = nodes.map { it.position }
+        val sortedSplits = splits.map { it.toSplitWay(positions) }.sorted()
+
+        val uploadElements = mutableListOf<Element>()
+        var newNodeId = -1L
```

Comment:
Always start from -1? What happens with more than one way split in changeset (or is it guaranteed to not happen)?

Reply:
Yes, always start from -1. The "temporary ids" (negative ids) are, to my knowledge, scoped per POST call to `changeset/$id/upload`, not per changeset. This makes sense, if you think about that the newly added elements will be on the map immediately after that call, not only after the changeset is closed.